src/state.rs

```diff
@@ -72,6 +73,46 @@ pub struct ProxyState {
 
     #[serde(flatten)]
     pub policies: PolicyStore,
+
+    pub resolved_dns: ResolvedDnsStore,
+}
+
+/// A ResolvedDnsStore encapsulates all per-workload resolved DNS information for workloads in the mesh
+#[derive(serde::Serialize, Default, Debug)]
+pub struct ResolvedDnsStore {
+    // workload UID to resolved IP addresses
+    by_workload_uid: HashMap<String, HashSet<IpAddr>>,
+}
+
+impl ResolvedDnsStore {
+    pub fn load_balance_for_workload(&self, workload_uid: String) -> Result<IpAddr, Error> {
+        debug!(
+            "load_balance_for_workload({}) by_workload_uid len: {}",
+            &workload_uid,
+            self.by_workload_uid.len()
+        );
+        // TODO: add support for different async dns modes: e.g. logical DNS vs strict DNS.
+        // this is effectively envoy strict DNS; returning a pre-resolved IP address at random.
+        let Some(ipset) = self.by_workload_uid.get(&workload_uid) else {
+            return Err(Error::NoResolvedAddresses(workload_uid));
+        };
+        let ips_vec = ipset.iter().collect::<Vec<_>>();
```

Comment:
I think you can choose() without collecting

Reply:
addressed